### PR TITLE
feat: adding cache and sync functionality

### PR DESCRIPTION
**Pull request by wei3erHase**

### CODE REVIEW

File: test/e2e/data-receiver.spec.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ describe('@skip-on-coverage DataReceiver.sol', () => {
 
     context('when an oracle is registered', () => {
       beforeEach(async () => {
-        await dataReceiver.addObservations(observationsData, salt, nonce - 1);
+        await dataReceiver.addObservations([[0, 0]] as unknown as IOracleSidechain.ObservationDataStruct[], salt, nonce - 1);
```
Comment (0xJabberwock) [inline test/e2e/data-receiver.spec.ts:73]: Is casting to `unknown` first required?

File: test/unit/DataFeed.spec.ts [hunk 2]
```diff
@@ -258,18 +258,12 @@ describe('DataFeed.sol', () => {
...
-            observationsData = [observationData1, observationData2];
-
             const tx = await dataFeed.connect(strategy).fetchObservations(randomSalt, secondsAgos);
             let eventPoolSalt = await readArgFromEvent(tx, 'PoolObserved', '_poolSalt');
             let eventPoolNonce = await readArgFromEvent(tx, 'PoolObserved', '_poolNonce');
-            let eventObservationsData = await readArgFromEvent(tx, 'PoolObserved', '_observationsData');
 
             expect(eventPoolSalt).to.eq(randomSalt);
             expect(eventPoolNonce).to.eq(nonce);
-            expect(eventObservationsData).to.eql(observationsData);
```
Comment (0xJabberwock) [inline test/unit/DataFeed.spec.ts:272]: Has the deleted expectation stopped working?

**Review by 0xJabberwock — APPROVED** [review]
Comments are 🔥!

File: test/e2e/data-receiver.spec.ts (hunk 1 above)
Comment (wei3erHase) [reply test/e2e/data-receiver.spec.ts:73]: yup :cries in ts:

File: test/unit/DataFeed.spec.ts (hunk 2 above)
Comment (wei3erHase) [reply test/unit/DataFeed.spec.ts:272]: stopped existing!

Comment (0xJabberwock) [reply test/unit/DataFeed.spec.ts:272]: What do you mean? `PoolObserved` event still emits `_observationsData`.

File: solidity/contracts/DataReceiver.sol [hunk 3]
```diff
@@ -46,9 +48,36 @@ contract DataReceiver is IDataReceiver, Governable {
...
+      // Query pool's current nonce
+      uint24 _currentNonce = _oracle.poolNonce();
+      // Discard old observations (already written in the oracle)
+      // NOTE: if _currentNonce == _poolNonce it shouldn't reach this else block
+      if (_currentNonce > _poolNonce) revert ObservationsNotWritable();
+      // Store not-added observations to cachedObservations mapping
+      // NOTE: memory to storage is not supported
+      // cachedObservations[_poolSalt][_poolNonce] = _observationsData;
+      for (uint256 _i; _i < _observationsData.length; ++_i) {
+        _cachedObservations[_poolSalt][_poolNonce].push(_observationsData[_i]);
```
Comment (0xJabberwock) [inline solidity/contracts/DataReceiver.sol:62]: Beware! The behavior of `_cachedObservations[_poolSalt][_poolNonce] = _observationsData` is not equal to `_cachedObservations[_poolSalt][_poolNonce].push(_observationsData[_i])`. The assignment would override the storage array but `push()` adds to it.

**Review by 0xJabberwock — CHANGES_REQUESTED** [review]

File: test/unit/DataFeed.spec.ts (hunk 2 above)
Comment (wei3erHase) [reply test/unit/DataFeed.spec.ts:272]: oh!!

File: solidity/contracts/DataReceiver.sol (hunk 3 above)
Comment (wei3erHase) [reply solidity/contracts/DataReceiver.sol]: great catch!

File: test/e2e/data-receiver.spec.ts [hunk 4]
```diff
@@ -122,4 +167,77 @@ describe('@skip-on-coverage DataReceiver.sol', () => {
...
+        beforeEach(async () => {
+          await dataReceiver.addObservations([[0, 0]] as IOracleSidechain.ObservationDataStructOutput[], salt, nonce + 2);
+          await dataReceiver.addObservations([[1, 0]] as IOracleSidechain.ObservationDataStructOutput[], salt, nonce + 1);
+          await dataReceiver.addObservations([[2, 0]] as IOracleSidechain.ObservationDataStructOutput[], salt, nonce);
+          await dataReceiver.addObservations([[3, 0]] as IOracleSidechain.ObservationDataStructOutput[], salt, nonce - 1);
+          // NOTE: dataReceiver should be at poolNonce == nonce by now (with nonce, nonce+1 & nonce+2 in cache)
+        });
+
+        it('should all the observations when called without max', async () => {
+          tx = await dataReceiver.syncObservations(salt, 0);
```
Comment (0xJabberwock) [inline test/e2e/data-receiver.spec.ts:215]: `'should add the observations...'`

**Review by 0xJabberwock — APPROVED** [review]
👍

File: solidity/interfaces/IDataReceiver.sol [hunk 5]
```diff
@@ -61,6 +61,12 @@ interface IDataReceiver is IGovernable {
     uint24 _poolNonce
   ) external;
 
+  /// @notice Allows any address to attemt to insert cached observations
```
Comment (0xJabberwock) [inline solidity/interfaces/IDataReceiver.sol:64]: `Allows any address to attempt to...`

**Pull request merged**